### ISSUE TITLE
RC2 configuration

### ISSUE THREAD

**Issue by alexeyignatiev** [logic-and-learning-lab/Popper#87]
Hi there!

As far as I can see in the README, you say that RC2 is much worse than a few other MaxSAT solvers. Although it may well be the case in general, I would recommend using a better configuration of RC2 because the default one (the one you use) is not designed to be fast. In particular, you may want to enable one of the competition versions, which perform much better than the baseline. 🙂 There are a number of parameters you can set depending on your needs. I can help you select the right one.

Best,
Alexey

**Comment by andrewcropper** [comment]
Hi Alexey,

Thanks for letting us know. 

Yes, please suggest some parameters. It would help us a lot to improve performance. 

What information would help you?

Thanks,

Andrew

**Comment by alexeyignatiev** [comment]
OK, do you compute a single optimal solution or do you enumerate them? Also, do your MaxSAT formulas are weighted or unweighted?

**Comment by andrewcropper** [comment]
A single optimal solution.

Weighted formula. 

Thanks!

**Comment by alexeyignatiev** [comment]
OK, then you may want to try configuration `B`. For that, you can create an object of the solver as follows:
```python
rc2 = RC2Stratified(formula, solver='g3', adapt=True, exhaust=True, blo='div', incr=False, minz=True, trim=0)
```
This assumes you first create your `WCNF` formula given as input to the solver. (This way it will process the formula in one go, which should be more efficient.)

You can also try another value for parameter `blo='cluster'` (it may work better in some cases).

**Comment by andrewcropper** [comment]
I just tried that change on a few sample instances. Below are the solving/termination times (seconds) for the old (default) and the new configuration:

```
Old                 New
------------------------------
1                   10
2                   0
4                   0
787                 0
1200 (timeout)      1
1200 (timeout)      14
```

That is an amazing reduction in solving time on the harder instances. Thanks a lot!

I also quickly compared RC2 with the new configuration against UWrMaxSat. UWrMaxSat is still a little faster, but not massively so. 

I will push this change soon and change the language in the Popper readme file.

Thanks again for the advice.

Kind regards,

Andrew

**Comment by yangdinglou** [comment]
@andrewcropper My recent experiments remind me this. You used to mention that the use of Clingo+weak constraints is much slower than MaxSAT. Did you try `--opt-strategy=usc` or simply `-t8`? The unclasp-based optimisation can be much better than the default branch-and-bound strategy in Clingo.

**Comment by andrewcropper** [comment]
@yangdinglou yeah, we tried loads of different Clingo settings.

For optimisation problems, when given the same number of cores/threads, we found Clingo to be nowhere near almost all MaxSAT solvers. The differences were hours vs seconds.

I recently benchmarked various solvers on the combine stage of Popper (an optimisation problem) and ended up switching the default solver to CP-SAT: https://github.com/logic-and-learning-lab/Popper/blob/main/popper/combiner_mdl.py . I cannot remember the exact figures, but CP-SAT outperformed Clingo on every optimisation problem. CP-SAT is still quite a lot slower than MaxSAT solvers, but it is a bit easier to use in Popper because we get intermediate solutions, which I was struggling to do with RC2.
